torchtext/experimental/datasets/translation.py

```diff
@@ -0,0 +1,539 @@
+            TranslationDataset(raw_data[key], (src_vocab, tgt_vocab),
+                               (src_text_transform, tgt_text_transform)))
+
+    return tuple(datasets)
+
+
+class TranslationDataset(torch.utils.data.Dataset):
+    """Defines a dataset for translation.
+       Currently, we only support the following datasets:
+             - Multi30k
+             - WMT14
+             - IWSLT
+    """
+    def __init__(self, data, vocab, transforms):
+        """Initiate translation dataset.
+
+        Arguments:
+            data: a tuple of source and target tensors, which include token ids
+                numericalizing the string tokens.
+                [(src_tensor0, tgt_tensor0), (src_tensor1, tgt_tensor1)]
+            vocab: source and target Vocabulary object used for dataset.
+                (src_vocab, tgt_vocab)
+            transforms: a tuple of source and target string transforms.
+
+        Examples:
+            >>> from torchtext.vocab import build_vocab_from_iterator
+            >>> src_data = torch.Tensor([token_id_s1, token_id_s2,
+                                         token_id_s3, token_id_s1]).long()
+            >>> tgt_data = torch.Tensor([token_id_t1, token_id_t2,
+                                         token_id_t3, token_id_t1]).long()
+            >>> src_vocab = build_vocab_from_iterator([['Übersetzungsdatensatz']])
+            >>> tgt_vocab = build_vocab_from_iterator([['translation', 'dataset']])
+            >>> dataset = TranslationDataset([(src_data, tgt_data)],
+                                              (src_vocab, tgt_vocab))
+        """
+
+        super(TranslationDataset, self).__init__()
+        self.data = data
+        self.vocab = vocab
+        self.transforms = transforms
+
+    def __getitem__(self, i):
+        source = self.transforms[0](self.data[i][0])
+        target = self.transforms[1](self.data[i][1])
+        return (source, target)
+
+    def __len__(self):
+        return len(self.data)
+
+    def get_vocab(self):
+        return self.vocab
+
+
+def Multi30k(train_filenames=("train.de", "train.en"),
+             valid_filenames=("val.de", "val.en"),
+             test_filenames=("test_2016_flickr.de", "test_2016_flickr.en"),
+             tokenizer=None,
+             root='.data',
+             vocab=(None, None),
+             data_select=('train', 'valid', 'test'),
+             removed_tokens=['<unk>']):
+    """ Define translation datasets: Multi30k
+        Separately returns train/valid/test datasets as a tuple
+
+    Arguments:
+        train_filenames: the source and target filenames for training.
+            Default: ('train.de', 'train.en')
+        valid_filenames: the source and target filenames for valid.
+            Default: ('val.de', 'val.en')
+        test_filenames: the source and target filenames for test.
+            Default: ('test2016.de', 'test2016.en')
+        tokenizer: the tokenizer used to preprocess source and target raw text data.
```

Comment:
We should explicitly say that the tokenizer needs a tuple of tokenizers. And the default is (get_tokenizer("spacy", language='de_core_news_sm'), get_tokenizer("spacy", language='en_core_web_sm')). Same for vocab.

Reply:
done